### ISSUE TITLE
Suggestion: reduce the amount of types

### ISSUE THREAD

**Issue by Datseris** [pablosanjose/Elsa.jl#3]
Standard Julia already offers a lot of Types and power through multiple dispatch.

Why does this exist: `LinkRule(1/√3), LatticeConstant(0.246)` when the values themselves are enough to dispatch? Why can't we use instead `1/sqrt(3)` and `0.245`?

Not having all these extra types will significantly improve the API. A high level constructor can take `a` and make it `LatticeConstant(a)`.

Could also clear the code, why are they needed in the code? I can't understand why.

**Comment by pablosanjose** [comment]
This is a deliberate choice in Elsa.jl. The idea is that a `Lattice` is built by chaining instructions, dubbed `LatticeDirectives`, which include `Sublat`, `Bravais`, `LinkRule`, `Region`, `Supercell`, `LatticeConstant`, `Dim` and `Precision`.

**Comment by Datseris** [comment]
Can you justify why the order of providing these things matters? Let's strip it down to the super-minimal stuff: A lattice is a combination of a basis and a set of bravais vectors. Both of these are unique and independent. 

1. The lattice constant is part of the bravais vectors. Why need it be an independent variable? Is it not enough to say `a0 = 0.5; bravais = (a0 .* (1, 0), a0 .* (0, 1)) ` ? But even so, it can be a field of a `Bravais` object. So that `Bravais` is initialized with the first argument always `<: Real`. E.g. `Bravais(a0, v1, v2, ...)`. The inner constructor can even go as far as normalize `v1, v2`. I Find this an intuitive definition of the bravai vectors.

1. Why is precision argument necessary, when it can be deduced from the vectors used in the definition of Bravais? An even then, why isn't a simple numerical type enough like `Float64, BigFloat` but has to be wrapped in another argument yet again? Why would the order of this matter?

1. Why would you specify `Dim`? It is also deduced by the number of bravais vectors and their number is known at compile time because they are a tuple.

2. What is a supercell ? Isn't the unit cell infinite by itself?

2. Why is Region part of the lattice construction? Kwant separates the creation of a Tight Binding system from the creation of a lattice. The Lattice is an abstract thing, by definition always infinite. The System/Model covers some finite `Region`. Separating the definitions is advantageous. At the moment the line `Lattice(Sublat((0.0, -0.5/sqrt(3.0)); name = :A), Sublat((0.0, 0.5/sqrt(3.0)); name = :B), Bravais((cos(pi/3), sin(pi/3)), (-cos(pi/3), sin(pi/3))), LinkRule(1/√3), LatticeConstant(0.246))` is very confusing because it has way too much information.

I can see later on that there is the `Model` object that has hoppings. If there is already another object that provides hopping information, what does the `LinkRule` do?

## How Kwant does it

Here is how one creates any arbitrary lattice in kwant for any dimensions. There is the constructor `Lattice` which takes as first argument the bravais vectors and as second argument the basis vectors. For example:
```
graphene = kwant.lattice.general([(1, 0), (sin_30, cos_30)],
                                 [(0, 0), (0, 1 / sqrt(3))])
a, b = graphene.sublattices
```
(Of course the above information is enough to automatically deduce the sub lattices)

If you want a detached lattice constant you can make the constructor (in Julia)
```
function Lattice(a0::Real, bravais::Tuple{Union{SVector, NTuple}}, sublattices = ())
```
(in the default case there are no sublattices)

To create a finite system, you provide a `Region` function to a "Builder" object. The "Builder" object is the tight binding system, not the lattice.

```
syst = kwant.Builder()

    #### Define the scattering region. ####
    # Now, we aim for a more complex shape, namely a ring (or annulus)
    def ring(pos):
        (x, y) = pos
        rsq = x ** 2 + y ** 2
        return (r1 ** 2 < rsq < r2 ** 2)
```

the _actual_ system then is defined by giving values to onsite elements that exist within the region (above `ring`):
```
syst[lat.shape(ring, seed) = function_of_position # or constant
```
before that no actual lattice sites exist.

Identical process is followed for the hoppings. You can either write down the hoppings yourself, e.g.:
```
hoppings = (((0, 0), a, b), ((0, 1), a, b), ((-1, 1), a, b))
```
(here `a, b` are sublattices). Given a lattice object kwant also provides "hoppings" that couple nearest neighbors, next nearest, etc. Using the above syntax you add hoppings to your system. Before that there are no hoppings. 

### A point you depart from Kwant
In the way kwant works, the "lattice" **does not have hoppings**. I think at the current state when you make a lattice you also integrate the hoppings. I think there is merit for the hopings to be added afterwards: it allows for space-dependent hoppings. E.g. if you have a magnetic field the value of the hopping depends on the position in space. You can only have the position in space only after you have a finite system, i.e. added a `Region`.

# Conclusion

There is no reason to follow the approach of Kwant. However kwant shows that you can make a tight binding system with a minimum amount of arguments, which are:
1. Set of bravais vectors (including lattice constant). Can make this a struct `Bravais` with the l.c. as a dedicated field.
2. Set of basis vectors (which can default to `(0, 0)`.

The above make the abstract lattice type. I find it intuitive that the lattice is detached from the tight binding system. Then you have:

3. A `region` boolean function. Defines the actual instance of the tight binding system
4. An onsite function: defines what the latice site values of the hamiltonian actually are. A lattice site exists only when the Hamiltonian obtains a value for it. (notice that this is dictated by the `region` function`).
5. A "hopping directive" (or HoppingKind): something that specifies which lattice sites couple to which.
6. A hopping function: given element (i, j) of the Hamiltonian matrix, it returns the entry the matrix should have. Number 5 makes number 6 easy.

---

P.s.: Could you please explain what the word "directive" means in your perspective and in the documentation? Is a "directive" an argument? To this point I have not understood what it is. It can be an instruction, because a number (`LatticeConstant(0.5)` ) is not an instruction but a number.

**Comment by pablosanjose** [comment]
> Can you justify why the order of providing these things matters? 

The updated quick-start guide has now an example. The basic idea of the API is that different instructions compose in an order dependent way. This is by design, and in my view allows for much greater flexibility than just having orthogonal commutative properties to define a lattice. One  possible example that I have encountered in real work is: take a graphene flake, linked to nearest neighbors, and find its spectrum after performing an elastic deformation. There are two possibilities: link the lattice first and then deform it without changing it's connectivity, or deform it, and then connect sites to their neighbors. The two are different, and both might be interesting in different contexts. This is just one example.

> Why is precision argument necessary, when it can be deduced from the vectors used in the definition of Bravais?

It is not necessary. If not specified it is indeed deduced from the definition of Bravais. The same applies to the embedding dimensions. The directives `Dim` and `Precision` *can* be used to override them, however. There are thus several ways to achieve the same. For example supply a `Bravais` with `Float64` coordinates and the supply a `Precision(Float32)` *or* do it as you suggest, writing the `Bravais` with `Float32`s.

> Why would you specify Dim?

See above.

> What is a supercell ? Isn't the unit cell infinite by itself?

A supercell is a larger unit cell for a Bravais lattice. There are many ways to choose the unit cell of a given lattice. Different choices allow different models to be applied to the Bravais lattice. For example, a square Bravais lattice with a 2x2 unit cell allows one to build a periodic system with a staggered potential, which would not be possible with a unit cell of a single atom.

> Why is Region part of the lattice construction?

I distinguish bounded lattices from infinite lattices. They differ in their lattice dimension `L` (parameter of `Lattice{T,E,L}`). A bounded lattice has `L=0`. `Region` allows one to create a bounded lattice with arbitrary shape from an unbounded lattice, used to fill the region. See updated docs for more detail.

> I can see later on that there is the Model object that has hoppings. If there is already another object that provides hopping information, what does the LinkRule do?

`LinkRule` specifies *geometric* information, i.e. it defines which sites are coupled to which others sites. It doesn't say anything about the physical model, the hopping amplitudes, the number of orbitals per site and the like. That is what `Model` does.

This design separates the geometrical object `Lattice` from the physical object `System`, which knows about the Hamiltonian. In a way, `Lattice` is just a rather general type of graph, with sublattices and Bravais vectors. Things can be done differently, but I have found this design to be extremely flexible. For example I can (and do) use a `Lattice` object to do a finite-element discretization of the Brillouin zone for bandstructure calculations. I thus make use of all its functionality in a different context.

> A point you depart from Kwant

I hope the above discussion clarifies the design better. If you still feel there is some situation in which the Kwant API allows more compact but still sufficiently flexible building, let me know, I'd be very interested. I've really thought long and hard about the current API.

**Comment by pablosanjose** [comment]
One point where the nomenclature can be different from kwant is the idea of `sublattice`. In Elsa.jl a sublattice is a subset of sites in the unit cell. Does kwant ground sites into sublattices? I think pybinding equates site in a unit cell with sublattice

**Comment by pablosanjose** [comment]
Another possible point of confusion: one can implement position-dependent onsite energies and hoppings very simply when building a model (which includes magnetic fields, disorder, etc). Just provide a function such as `Onsite(r-> cos[r[1]])` or `Hopping((r, dr) -> exp(im*dot(A(r), dr))` for a Peierls phase

**Comment by Datseris** [comment]
So far up to the `Supercell` part you offer proper justifications. Whether they are the best possible design will be discussed further as the package gains more use. Regardless, they are all justified properly.

---

Next, on the part about region:

> >  Why is Region part of the lattice construction?

> I distinguish bounded lattices from infinite lattices. They differ in their lattice dimension L (parameter of Lattice{T,E,L}). A bounded lattice has L=0. Region allows one to create a bounded lattice with arbitrary shape from an unbounded lattice, used to fill the region. See updated docs for more detail.

I found this extremely untintuitive: A bounded lattice has 0 dimension? In my eyes any 2D lattice has 2 dimensions regardless of size. 

This is difficult to get across to a newcomer and must be justified more. I don't understand: what is the reason of having a "finite lattice"? When creating the model the `Region` & `Onsite` argument functions can be used there. _They_ can specify the finite system. How can a lattice be finite (from a mathematics perspective)?

Why would a finite lattice be necessary _beforehand_? What is the advantage?

---

> LinkRule specifies geometric information, i.e. it defines which sites are coupled to which others sites. It doesn't say anything about the physical model, the hopping amplitudes, the number of orbitals per site and the like. That is what Model does.

Alright, but then why? Why is this necessary and what does it offer? `Model` already does this since it provides the physical information and thus _also_ the geometric information, as any hopping with physical value immediatelly obtains geometric connection as well.

I think the following may justify it. Correct me if I am wrong:

In your design, 

1. LinkRule(distance) connects all sites within a distance in an abstract way. "Geometrically" as you call it.
2. The _user_ provided function `Hopping((r, dr) -> exp(im*dot(A(r), dr))` can choose whether to give the value `t` if `norm(dr) == a0/sqrt(3)` or the value `tprime` if `norm(dr) == a0`. 
3. All sites are connected abstractly and therefore all possible connections are given to `Hopping`.

Then, I believe that this API must be somehow improved. It is much easier to do what kwant does: 
```
gsys[glat.neighbors()] = nn_hopping
gsys[glat.neighbors(2)] = nnn_hopping
```
where the second line automatically applies to next-nearest-neighbor while the first to only nearest neighbors. Notice that in the above design the `nn_hopping` are also functions that are given as arguments the two lattice positions (I believe the `r, dr` argument design is better though).

I guess what they do in kwant is that once you specify the lattice it is known "geometrically" which connections apply to which neighbors. In a sense, you don't have to provide `LinkRule` because they automatically generate the hopping generator `neighbors(n)` which lists hopping between `n`-nearest neighbors.

---

> This design separates the geometrical object Lattice from the physical object System, which knows about the Hamiltonian. In a way, Lattice is just a rather general type of graph, with sublattices and Bravais vectors. 

This is identical with what Kwant does. But your Lattice has much, much more information than Kwant's while your `System` has exactly the same amount of information as Kwants: 2 functions, one for the hoppings the other for the onsite energy. (In kwant the `Region` function is part of the `System` and not the `Lattice`).

I guess your second part of the answer tells me why your Lattice design needs so much more information: 

> For example I can (and do) use a Lattice object to do a finite-element discretization of the Brillouin zone for bandstructure calculations. I thus make use of all its functionality in a different context.

although how can one do band structure calculations without the hopping energy information still baffles me.

**Comment by Datseris** [comment]
A big difference: In Kwant, all lattice information is just numbers and vectors. In Elsa everything is wrapped in a dedicated type.

This is done so that this "sequential directive" method applies, but I believe a high level constructor
```
lattice(a, bravais, linkrule; basis = zero(bravais[1]))
```
which accepts numbers and vectors is necessary. New users don't want to immediatelly see these highly advanced ways to construct a lattice.

**Comment by pablosanjose** [comment]
Ok! That can be done. No need for the a, though, that would be derived from bravais. The change would amount to adding a new exported function such as 
```
lattice(bravais::NTuple{L, NTuple{E, <:Real}}, linkrange::Real; basis =  zero.(bravais[1])) where {L,E} = Lattice(Sublat(basis), Bravais(bravais), LinkRule(linkrange))
```

Though I'm not so sure what is gained by this. In the user's code I think this
```
Lattice(Sublat(r1), Bravais(a1, a2), LinkRule(range))
```
is clearer (more explicit) than
```
lattice((a1, a2), range, basis = r1)
```

**Comment by pablosanjose** [comment]
> I found this extremely untintuitive: A bounded lattice has 0 dimension? In my eyes any 2D lattice has 2 dimensions regardless of size.

Yes. The lattice dimension is by (my) definition the number of Bravais vectors. Any bunch of sites in space that are not periodically copied in any direction define a zero-dimensional lattice (no Bravais vectors). In this sense, a bounded graphene flake is not a 2D lattice, but a zero-dimensional lattice. 

You might be confused by this because you are still thinking of the graphene flake as a repetition of two-carbon-atoms unit cell, even though they don't fill up space. However consider that you add some strong positional disorder to the flake. You can no longer indentify an underlying lattice. If you have a chunk of amorphous carbon of finite size, is that a 3D lattice? For me it is a finite collection of carbon sites, not repeated in space, hence a 0D lattice.

The need to define large bounded lattices in kwant or Elsa is obvious: to describe the effect of confinement, boundaries, arbitrary (inconmensurate) magnetic fields, edge states, topological boundary modes, etc.

> Then, I believe that this API must be somehow improved. It is much easier to do what kwant does

As far as I undersrtand from your post, Kwant specifies the hopping to a given distance as `gsys[glat.neighbors(2)] = nnn_hopping`. In Elsa you specify hopping by `(r, dr)` and by the sublattices of the two sites. In some case the former is easier (e.g. the Haldane model), while in others the latter is easier (twisted graphene bilayer). I understand the `2` above does not refer to a distance in real space but rather a metropolis distance in unit cells? Perhaps the actionable issue here would be to extend the `Hopping` functionality to specify this.

**Comment by pablosanjose** [comment]
> This is identical with what Kwant does. But your Lattice has much, much more (actually it is 3 arguments more, it is not much much more) information than Kwant's while your System has exactly the same amount of information as Kwants: 2 functions, one for the hoppings the other for the onsite energy. (In kwant the Region function is part of the System and not the Lattice).

Perhaps you can write down a specific example in kwant and Elsa side by side and try to argue why kwant is preferrable. I might be able to argue the opposite :-)

**Comment by pablosanjose** [comment]
> In your design,
LinkRule(distance) connects all sites within a distance in an abstract way. "Geometrically" as you call it.
The user provided function Hopping((r, dr) -> exp(im*dot(A(r), dr)) can choose whether to give the value t if norm(dr) == a0/sqrt(3) or the value tprime if norm(dr) == a0.
All sites are connected abstractly and therefore all possible connections are given to Hopping

Maybe there is a misunderstanding: only sites that are linked in a `lattice` get a hopping in `System(lattice,Model(...))`. The rest of sites don't get a hopping. Did I misundertand what you meant?

**Comment by pablosanjose** [comment]
I've been thinking more about your comments, and decided I'd like to play a bit with the API and internal structure to see how something closer to pybinding and kwant could work. Hopefully it will relieve some of your concerns. It's not a drastic change, so I don't expect it to be so much work, but please be patient. The changes, moreover, might be important for lazy evaluation of hamiltonians in the future (matrix-free methods). I'll post an experimental PR soon, time permitting.

**Comment by pablosanjose** [comment]
Closed by #6

The new API design remains substantially different from Kwant, but is now free of some limitations brought to light by this issue. The code is simpler and quite a bit faster now than before, and uses less memory. Internally, we've reworked the links machinery, which is now purely based on `SparseMatrixCSC` Hamiltonians and their manipulations. This required a very substantial rewrite of build algorithms. As a result current master is not yet at feature-parity with the previous version 0.1.
